content/article/json-unicode-and-perl-oh-my-.md

```diff
@@ -0,0 +1,325 @@
+see something like this for Mojo::JSON’s output:
+```
+SV = PV(0x7fdc27802f30) at 0x7fdc27e59c58
+  REFCNT = 1
+  FLAGS = (POK,pPOK)
+  PV = 0x7fdc28826350 "[\"\303\203\302\251\"]"\0
+  CUR = 8
+  LEN = 34
+```
+… and this for Cpanel::JSON::XS:
+```
+SV = PV(0x7fc0cd004d30) at 0x7fc0cd016228
+  REFCNT = 1
+  FLAGS = (POK,pPOK,UTF8)
+  PV = 0x7fc0cce2ef60 "[\"\303\203\302\251\"]"\0 [UTF8 "["\x{c3}\x{a9}"]"]
+  CUR = 8
+  LEN = 34
+```
+Note the `UTF8` flag in the latter. This tells us that Perl’s
+internal storage of the string’s code points uses UTF-8 encoding. This
+difference is why, as we saw earlier, Data::Dumper encodes Mojo::JSON’s output
+using
+octal escapes but Cpanel::JSON::XS’s using hex: Data::Dumper recognizes the
+UTF8 flag and renders its output based on it.
+
+As (perldoc perlunifaq)[https://perldoc.perl.org/perlunifaq.html#What-is-%22the-UTF8-flag%22%3f] makes clear, the UTF8 flag is _not_ meant for consumption
+by Perl code. Nevertheless, it often does work to regard UTF8-flagged
+strings as “character strings” and non-UTF8-flagged strings as “byte
+strings”—indeed, multiple
+serializers on CPAN, including two of my own, do exactly this. This isn’t
+a supported model, though, for understanding Perl strings, and any code that
+depends on it may behave differently in different Perl versions. Caveat
+emptor!
+
+Making Peace
+============
+
+JSON and Perl are odd bedfellows. Perl’s lack of distinct number and string
+types, for example, can yield JSON that uses the wrong type for one
+value or the other. Perl’s lack of native booleans produces a similar effect.
+
+The encoding problems discussed above, though, are particularly nefarious
+because correcting for them requires a deeper understanding of all of this.
+For example, most developers can accommodate `{"age": "9"}` easily enough
+because typecasting from `"9"` (string) to `9` (number) is commonplace. But
+how many would see `"Ã©"` and think, “ah! I simply have to treat those
+characters’ code points as bytes then decode those bytes as UTF-8!” Some
+would, to be sure—perhaps even many—but likely fewer than can easily coerce
+`"9"` to `9`.
+
+Binary-friendly encodings like [CBOR](http://cbor.io)
+mitigate against this problem because whatever decodes the Perl-sourced
+data can more easily recognize the need to decode from binary. Anyone
+who doesn’t know about bytes and encodings will quickly learn! Fundamentally,
+though, even CBOR doesn’t really fit Perl’s “pure code points” string model
+very well because CBOR distinguishes strongly between binary and text strings,
+which Perl does not.
+
+At the end of the day, Perl’s data model, for all of the conveniences that it
+affords us, makes communication with other modern languages a challenge. The
+best we can do is to know of these problems and deal with them as they arise.
+
+Epilogue: JSON Alternatives
+===========================
+
+Being restricted to Unicode is, in my opinion, JSON’s biggest drawback,
```

Comment:
Did you mean UTF-8 instead of "Unicode" here?

Reply:
Actually, I did mean Unicode, though I’ll reword, as I can see why it’s unclear.